content/en/blog/_posts/2024-05-01-cri-streaming-explained.md

```diff
@@ -0,0 +1,338 @@
+    )
+}
+```
+
+### **PortForward**
+
+Forwarding ports to a container works a bit differently when comparing it to
+streaming IO data from a workload. The server still has to provide a URL
+endpoint for the client to connect to, but then the container runtime has to
+enter the network namespace of the container, allocate the port as well as
+stream the data back and forth. There is no simple protocol definition available
+like for **Exec** or **Attach**. This means that the client will stream the
+plain SPDY frames (with or without an additional WebSocket connection) which can
+be interpreted using libraries like [moby/spdystream](https://github.com/moby/spdystream).
+
+Luckily, the kubelet library already provides the `PortForward` interface method
+which has to be implemented by the runtime. [CRI-O does that]() by (simplified):
+
+```go
+func (s StreamService) PortForward(
+    ctx context.Context,
+    podSandboxID string,
+    port int32,
+    stream io.ReadWriteCloser,
+) error {
+    // Retrieve the pod sandbox by the provided podSandboxID
+    sandboxID, err := s.runtimeServer.PodIDIndex().Get(podSandboxID)
+    sb := s.runtimeServer.GetSandbox(sandboxID)
+    // …
+
+    // Get the network namespace path on disk for that sandbox
+    netNsPath := sb.NetNsPath()
+    // …
+
+    // Enter the network namespace and stream the data
+    return s.runtimeServer.Runtime().PortForwardContainer(
+        ctx, sb.InfraContainer(), netNsPath, port, stream,
+    )
+}
+```
+
+## Future work
+
+The flexibility Kubernetes provides for the RPCs **Exec**, **Attach** and
+**PortForward** is truly outstanding compared to other methods. Nevertheless,
+container runtimes have to keep up with the latest and greatest implementations
+to support those features in a meaningful way. The general effort to support
+WebSockets is not only a plain Kubernetes thing, it also has to be supported by
+container runtimes as well as clients like `crictl`.
+
+For example, `crictl` v1.30 features a new `--transport` flag for the
+subcommands `exec`, `attach` and `port-forward`
+([#1383](https://github.com/kubernetes-sigs/cri-tools/pull/1383),
+[#1385](https://github.com/kubernetes-sigs/cri-tools/pull/1385))
+to allow choosing between `websocket` and `spdy`.
+
+CRI-O is going an experimental path by moving the streaming server
+implementation into [conmon-rs](https://github.com/containers/conmon-rs)
+(a substitute for the container monitor [conmon](https://github.com/containers/conmon)). conmon-rs is
+a [Rust](https://www.rust-lang.org) implementation of the original container
+monitor and allows streaming WebSockets directly using supported libraries
+([#2070](https://github.com/containers/conmon-rs/pull/2070)). The major benefit
+of this approach is that CRI-O does not even have to be running while conmon-rs
+can keep active **Exec**, **Attach** and **PortForward** sessions open. The
+simplified flow when using crictl directly will then look like this:
+
+{{< mermaid >}}
```

Comment:
nit: white text on blue background is easier to read.

If you've capacity, we'd welcome a follow-up PR.

Reply:
Sure, following-up in https://github.com/kubernetes/website/pull/46086